src/engine/TransitivePathBinSearch.cpp

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@neptun.uni-freiburg.de)
+
+#include "TransitivePathBinSearch.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "engine/CallFixedSize.h"
+#include "engine/TransitivePathBase.h"
+#include "util/Exception.h"
+#include "util/Timer.h"
+
+// _____________________________________________________________________________
+TransitivePathBinSearch::TransitivePathBinSearch(
+    QueryExecutionContext* qec, std::shared_ptr<QueryExecutionTree> child,
+    const TransitivePathSide& leftSide, const TransitivePathSide& rightSide,
+    size_t minDist, size_t maxDist)
+    : TransitivePathBase(qec, child, leftSide, rightSide, minDist, maxDist) {
+  auto [startSide, targetSide] = decideDirection();
+  subtree_ = QueryExecutionTree::createSortedTree(
+      subtree_, {startSide.subCol_, targetSide.subCol_});
+}
+
+// _____________________________________________________________________________
+template <size_t RES_WIDTH, size_t SUB_WIDTH, size_t SIDE_WIDTH>
+void TransitivePathBinSearch::computeTransitivePathBound(
+    IdTable* dynRes, const IdTable& dynSub, const TransitivePathSide& startSide,
+    const TransitivePathSide& targetSide, const IdTable& startSideTable) const {
+  auto timer = ad_utility::Timer(ad_utility::Timer::Stopped);
+  timer.start();
+
+  auto [edges, nodes] = setupMapAndNodes<SUB_WIDTH, SIDE_WIDTH>(
+      dynSub, startSide, targetSide, startSideTable);
+
+  timer.stop();
+  auto initTime = timer.msecs();
+  timer.start();
+
+  Map hull(allocator());
+  if (!targetSide.isVariable()) {
+    hull = transitiveHull(edges, nodes, std::get<Id>(targetSide.value_));
+  } else {
+    hull = transitiveHull(edges, nodes, std::nullopt);
+  }
+
+  timer.stop();
+  auto hullTime = timer.msecs();
+  timer.start();
+
+  fillTableWithHull(*dynRes, hull, nodes, startSide.outputCol_,
+                    targetSide.outputCol_, startSideTable,
+                    startSide.treeAndCol_.value().second);
+
+  timer.stop();
+  auto fillTime = timer.msecs();
+
+  auto& info = runtimeInfo();
+  info.addDetail("Initialization time", initTime.count());
+  info.addDetail("Hull time", hullTime.count());
+  info.addDetail("IdTable fill time", fillTime.count());
+}
```

Comment:
Isn't this exactly the same for the binSearch and the fallback with the difference that types are different, but the function calls are the same?
In this case you can also factor this out using e.g. a CRTP style Mixin to reduce the code duplication.

Reply:
(I think that this is the same for several other functions below).